File: intreview_training/algorithms/data-structures/implementations/python/hash_table.py
"""
Hash Table Implementation in Python
Key-value pairs with hash function
Time Complexity: O(1) average, O(n) worst
"""

class HashTable:
    def __init__(self, initial_capacity=16, load_factor=0.75):
        self.buckets = [None] * initial_capacity
        self.capacity = initial_capacity
        self.size = 0
        self.load_factor = load_factor

    def _hash(self, key):
        """Hash function - converts key to index"""
        hash_value = 0
        key_string = str(key)
        
        for char in key_string:
            hash_value = ((hash_value << 5) - hash_value) + ord(char)
            hash_value = hash_value & 0xFFFFFFFF  # Convert to 32-bit integer
        
        return abs(hash_value % self.capacity)

    def set(self, key, value):
        """Insert key-value pair - O(1) average"""
        index = self._hash(key)
        
        if self.buckets[index] is None:
            self.buckets[index] = []

        # Check if key already exists
        bucket = self.buckets[index]
        for i in range(len(bucket)):
            if bucket[i][0] == key:
                bucket[i][1] = value  # Update existing
                return self

        # Add new key-value pair
        bucket.append([key, value])
        self.size += 1

        # Resize if load factor exceeded
        if self.size / self.capacity > self.load_factor:
            self._resize()

        return self

    def get(self, key):
        """Get value by key - O(1) average"""
        index = self._hash(key)
        bucket = self.buckets[index]

        if bucket is None:
            return None

        for item in bucket:
            if item[0] == key:
                return item[1]

        return None

    def delete(self, key):
        """Delete key-value pair - O(1) average"""
        index = self._hash(key)
        bucket = self.buckets[index]

        if bucket is None:
            return False

        for i in range(len(bucket)):
            if bucket[i][0] == key:
                bucket.pop(i)
                self.size -= 1
                
                if len(bucket) == 0:
                    self.buckets[index] = None
                
                return True

        return False

    def has(self, key):
        """Check if key exists - O(1) average"""
        return self.get(key) is not None

    def keys(self):
        """Get all keys - O(n)"""
        keys_list = []
        for bucket in self.buckets:
            if bucket:
                for item in bucket:
                    keys_list.append(item[0])
        return keys_list

    def values(self):
        """Get all values - O(n)"""
        values_list = []
        for bucket in self.buckets:
            if bucket:
                for item in bucket:
                    values_list.append(item[1])
        return values_list

    def entries(self):
        """Get all entries - O(n)"""
        entries_list = []
        for bucket in self.buckets:
            if bucket:
                for item in bucket:
                    entries_list.append((item[0], item[1]))
        return entries_list

    def clear(self):
        """Clear hash table - O(n)"""
        self.buckets = [None] * self.capacity
        self.size = 0

    def _resize(self):
        """Resize hash table when load factor exceeded"""
        old_buckets = self.buckets
        self.capacity *= 2
        self.buckets = [None] * self.capacity
        self.size = 0

        # Rehash all entries
        for bucket in old_buckets:
            if bucket:
                for item in bucket:
                    self.set(item[0], item[1])

    def get_size(self):
        """Get size - O(1)"""
        return self.size

    def is_empty(self):
        """Check if empty - O(1)"""
        return self.size == 0


# Usage Example
if __name__ == "__main__":
    hash_table = HashTable()

    hash_table.set('name', 'John')
    hash_table.set('age', 30)
    hash_table.set('city', 'New York')
    hash_table.set('email', 'john@example.com')

    print('Get name:', hash_table.get('name'))  # 'John'
    print('Has age:', hash_table.has('age'))  # True
    print('Keys:', hash_table.keys())
    print('Values:', hash_table.values())

    hash_table.delete('age')
    print('After delete age:', hash_table.has('age'))  # False
    print('Size:', hash_table.get_size())  # 3

File: intreview_training/algorithms/data-structures/implementations/java/HashTable.java
/**
 * Hash Table Implementation in Java
 * Key-value pairs with hash function
 * Time Complexity: O(1) average, O(n) worst
 */
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class HashTable {
    private List<List<Entry>> buckets;
    private int capacity;
    private int size;
    private double loadFactor;

    private static class Entry {
        String key;
        Integer value;

        Entry(String key, Integer value) {
            this.key = key;
            this.value = value;
        }
    }

    public HashTable(int initialCapacity, double loadFactor) {
        this.capacity = initialCapacity;
        this.loadFactor = loadFactor;
        this.buckets = new ArrayList<>(capacity);
        for (int i = 0; i < capacity; i++) {
            buckets.add(new LinkedList<>());
        }
        this.size = 0;
    }

    public HashTable() {
        this(16, 0.75);
    }

    private int hash(String key) {
        int hash = 0;
        for (char c : key.toCharArray()) {
            hash = ((hash << 5) - hash) + c;
            hash = hash & hash; // Convert to 32-bit integer
        }
        return Math.abs(hash % capacity);
    }

    public HashTable set(String key, Integer value) {
        int index = hash(key);
        List<Entry> bucket = buckets.get(index);

        // Check if key already exists
        for (Entry entry : bucket) {
            if (entry.key.equals(key)) {
                entry.value = value; // Update existing
                return this;
            }
        }

        // Add new key-value pair
        bucket.add(new Entry(key, value));
        size++;

        // Resize if load factor exceeded
        if ((double) size / capacity > loadFactor) {
            resize();
        }

        return this;
    }

    public Integer get(String key) {
        int index = hash(key);
        List<Entry> bucket = buckets.get(index);

        for (Entry entry : bucket) {
            if (entry.key.equals(key)) {
                return entry.value;
            }
        }

        return null;
    }

    public boolean delete(String key) {
        int index = hash(key);
        List<Entry> bucket = buckets.get(index);

        for (int i = 0; i < bucket.size(); i++) {
            if (bucket.get(i).key.equals(key)) {
                bucket.remove(i);
                size--;
                return true;
            }
        }

        return false;
    }

    public boolean has(String key) {
        return get(key) != null;
    }

    public List<String> keys() {
        List<String> keysList = new ArrayList<>();
        for (List<Entry> bucket : buckets) {
            for (Entry entry : bucket) {
                keysList.add(entry.key);
            }
        }
        return keysList;
    }

    public List<Integer> values() {
        List<Integer> valuesList = new ArrayList<>();
        for (List<Entry> bucket : buckets) {
            for (Entry entry : bucket) {
                valuesList.add(entry.value);
            }
        }
        return valuesList;
    }

    private void resize() {
        List<List<Entry>> oldBuckets = new ArrayList<>(buckets);
        capacity *= 2;
        buckets = new ArrayList<>(capacity);
        for (int i = 0; i < capacity; i++) {
            buckets.add(new LinkedList<>());
        }
        size = 0;

        // Rehash all entries
        for (List<Entry> bucket : oldBuckets) {
            for (Entry entry : bucket) {
                set(entry.key, entry.value);
            }
        }
    }

    public int getSize() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public void clear() {
        for (List<Entry> bucket : buckets) {
            bucket.clear();
        }
        size = 0;
    }

    // Usage Example
    public static void main(String[] args) {
        HashTable hashTable = new HashTable();

        hashTable.set("name", 1);
        hashTable.set("age", 30);
        hashTable.set("city", 100);
        System.out.println("Get name: " + hashTable.get("name"));
        System.out.println("Has age: " + hashTable.has("age"));
        System.out.println("Keys: " + hashTable.keys());
        System.out.println("Size: " + hashTable.getSize());
    }
}

